newton.go

```diff
@@ -0,0 +1,203 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package optimize
+
+import (
+	"errors"
+	"math"
+
+	"github.com/gonum/floats"
+	"github.com/gonum/matrix/mat64"
+)
+
+// Newton implements pure Newton's method without linesearch for unconstrained
+// minimization of functions that satisfy Gradient and Hessian interfaces.
+//
+// Newton's method iteratively forms a quadratic model to the objective
+// function f and tries to minimize this approximate model. It generates a
+// sequence of locations x_k by means of
+//  solve H_k d_k = -∇f_k for d_k,
+//  x_{k+1} = x_k + d_k,
+// where H_k is the Hessian matrix of f at x_k.
+//
+// Note that Newton's method is only locally convergent. This means that for
+// starting points close to a minimizer the sequence x_k converges rapidly with
+// a quadratic rate. However, away from a minimizer the quadratic approximation
+// may not be valid, the Hessian may not be positive definite and d_k may not
+// be a descent direction. In such case Newton fails with an error.
+//
+// For a globally convergent Hessian-based method, see ModifiedNewton. If the
+// Hessian matrix cannot be formed explicitly or if the computational cost of
+// its factorization is prohibitive, BFGS or L-BFGS methods can be used
+// instead.
```

Comment:
I'm not sure this should be in this package. The package is for general nonlinear minimization, and it is very easy for a general nonlinear function to have a non-positive-definite hessian, or to diverge in the iterative Newton procedure. This algorithm is useful somewhere, but I would rather have the Newton method in this package be one that can handle the general class of functions.